File: normalFormToLaTeX.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.StringTokenizer;
import java.util.Arrays;

//to-do
//ability to enter parameters 1, 4, 5 (size of cell, name of row player, name of col player) (?)
//move the bimatrix macro format to file instead of hard-coding (?)
//some error handling for mal-formed inputs?
//support for fractions (rational numbers) including 
//		change payoff type from string to Number? Rational (similar to gte)?
//conversion to .eps format
//computation of best responses for row and column player and surrounding the payoffs with \fbox
public class normalFormToLaTeX 
{
	private ArrayList<ArrayList<String>> a1;
	private ArrayList<ArrayList<String>> a2;
	private String filename;
	private Boolean bSinglePayoff = false;
	
	/* constructor - by file which defines matrices */
	public normalFormToLaTeX(String fn)
	{
		filename = fn;
		splitMatrices(fileToArrayList());
	}
	
	/* constructor - payoff matrices passed in as parameters */
	public normalFormToLaTeX(String[][] p1, String[][] p2 )
	{
		filename = "bm_latex";
		//convert input strings to arraylist since that's what we're working with in this class
		a1 = arrayToArrayList2D(p1);
		
		if (p2 !=null ) { a2 = arrayToArrayList2D(p2); }
		else
		{
			bSinglePayoff = true;
			a2 = a1;
		}
	}
	
	public void setFilename(String fn)
	{
		filename = fn;
	}
	
	//reads the input file 
	//outputs an ArrayList containing the list of lines (strings) that make up the file
	public ArrayList<String> fileToArrayList() 
	{
		ArrayList<String> fileLines = new ArrayList();
		String fileLine;
		
		try
        {
			BufferedReader reader = new BufferedReader(new FileReader(filename));	
			
			//read all the lines from the file
			while ((fileLine = reader.readLine()) != null) 
			{
				fileLines.add(fileLine);
		    }
		    reader.close();
        }
		catch (Exception e)
		{
			//raise error regarding unsuccessfully reading the file
		}
		
		return fileLines;
	}
	
	//there may be either one or two matrices in the input text
	//if one, duplicate to make 2 symmetric matrices
	//otherwise output two as expected
	public void splitMatrices(ArrayList<String> fileLines)
	{
		a1 = new ArrayList();
		a2 = new ArrayList();
		boolean bSwitch = false;
		
		//loop through all the lines
		//putting them either into lists a1 or a2
		for (Iterator<String> i = fileLines.iterator(); i.hasNext();) 
		{
			String fileLine = (String)i.next();
			
			if (bSwitch) 
			{
				a2.add(this.textToArray(fileLine));
			}
			else if (fileLine.trim().length() == 0)  //check for blank dividing line
			{
				bSwitch = true;
			}
			else
			{
				a1.add(this.textToArray(fileLine));
			}
		}
		
		if (a2.size() < 1)
		{
			//copy array 1 to array 2 since there was no array 2 in the file
			//clone is a shallow copy; don't need deep copy here
			a2 = (ArrayList<ArrayList<String>>)a1.clone();
			bSinglePayoff = true;
		}
	}
	
	//take a line of text, tokenize, and return as ArrayList
	public ArrayList<String> textToArray(String t)
	{
		StringTokenizer st = new StringTokenizer(t);
		ArrayList<String> row = new ArrayList<String>();
		
  	    while (st.hasMoreTokens() ) 
  	    {	
  	    	row.add(st.nextToken());
  	    } 
  	    return row;
	}
	
	//map tokens found in the input file to the bimatrix macro template
	//improvement: read from file instead of hard-coding 
	public String mapTokensToBMTemplate()
	{
		//determine number of columns
		ArrayList<String> a1row;
		ArrayList<String> a2row;
		
		int numCols = a1.get(0).size();
		int numRows = a1.size();
		
		/* \bimatrixgame{4mm}{2}{3}{I}{II}%
	    {{T}{B}}%
	    {{l}{c}{r}}
	    {
	    \payoffpairs{1}{012}{421}
	    \payoffpairs{2}{301}{132}
	    } */
		
		String s ="\\bimatrixgame{4mm}";
		s = s + "{" + numRows + "}{" +numCols + "}{I}{II}%\n";
		
		//row labels & column labels
		s = s + this.generateRowLabels(numRows);
		s = s + this.generateColumnLabels(numCols);
		
		//payoff pairs
		s = s + this.generatePayoffPairs();

		return s;
	}
	
	public String getLaTeXString()
	{
		String s = "\\documentclass{article}\n\\input{gamesty}%\n\\begin{document} \n";
		s = s + this.mapTokensToBMTemplate();
		s =s + "\\end{document} \n";
		return s;
	}
	 
	public void createLaTeXFile()
	{
		try 
		{
		    BufferedWriter out = new BufferedWriter(new FileWriter(filename+".tex"));
		    out.write(""+ this.getLaTeXString());
		    out.close();
		} 
		catch (IOException e) 
		{
			//throw exception if can't print file
		}
	}
	
	public String generateRowLabels(int numRows) 
	{
		 String label;
		 char chr;
		 switch (numRows)
		 {
	         case 1:  label = "{{M}}%\n";		break;
	         case 2:  label = "{{T}{B}}%\n";	break;
	         case 3:  label = "{{T}{M}{B}}%\n";	break;
	         default: 
	         {
	        	 label = "{";
	        	 chr = 97; //97 = ascii 'a'
	        	 for (int i = 97; i < numRows+97; i++)
	        	 {
	        		 chr = (char) i;
	        		 label = label + "{" + chr + "}";
	        	 }
	        	 label = label + "}%\n";
	         }
		 }
	     return label;
	}
	
	public String generateColumnLabels(int numCols) 
	{
		String label;
		 char chr;
		 switch (numCols)
		 {
	         case 1:  label = "{{c}}%\n";		break;
	         case 2:  label = "{{l}{r}}%\n";	break;
	         case 3:  label = "{{l}{c}{r}}%\n";	break;
	         default: 
	         {
	        	 label = "{";
	        	 chr = 97; 			//97 = ascii 'a'
	        	 for (int i = 97; i < numCols+97; i++)
	        	 {
	        		 chr = (char) i;
	        		 label = label + "{" + chr + "}";
	        	 }
	        	 label = label + "}\n";
	         }
		 }
	     return label;
	}
	
	public String generatePayoffPairs() 
	{
		//a1 and a2 are the ArrayLists containing the payoff matrices
		/* {
	    \payoffpairs{1}{012}{421}
	    \payoffpairs{2}{301}{132}
	    } */
		String s="{\n";
		String r="";
		String strPayoff = "\\payoffpairs{";
		int numRows = a1.size();
		int numCols = a1.get(0).size(); 
		
		if (bSinglePayoff) {strPayoff = "\\singlepayoffs{"; }
		
		for(int i = 0; i < numRows; i++ )
		{
			r = strPayoff + (i+1) + "}{";
			
			for(int j = 0; j < numCols; j++) //player I payoffs
			{
				r = r+ "{$" + a1.get(i).get(j) + "$}";
			}
			r = r + "}" ;
			
			if (!bSinglePayoff)
			{
				r = r + "{";
				
				for(int j = 0; j < numCols; j++) //player II payoffs
				{
					r = r+ "{$" + a2.get(i).get(j) + "$}";
				}
				r = r + "}";
			}
			s = s + r + "\n";
		}
		
		s = s + "}\n";
		return s;
	}

	public ArrayList<ArrayList<String>> arrayToArrayList2D(String[][] s)
	{
		ArrayList<ArrayList<String>> al = new ArrayList<ArrayList<String>>();
		for (int i = 0; i < s.length; i ++)
		{
			ArrayList<String> tmp = new ArrayList<String>(Arrays.asList(s[i]));
			al.add(tmp);
		}
		
		return al;
	}
	
	public static void main (String [] args)
	{	
		String filename = args[0];
		normalFormToLaTeX ml = new normalFormToLaTeX(filename);
		ml.createLaTeXFile(); 
	}
}

File: normalFormToLaTeX.py
#last modified June 17, 2011

import sys
import getopt

class normalFormToLaTeX:
	""" Convert a matrix file to a LaTeX file."""
	def __init__(self,fname, formatFraction = True, showBestResponse = True): 
		""" Instantiate the class, parse the file, and determine best responses in preparation for formatting."""

		self.a1 = []
		self.a2 = []	
		
		self.numRows = 0		# number of rows in payoff matrices, will be updated
		self.numCols = 0		# number of columns in payoff matrices, will be updated
	
		self.bSinglePayoff = False 				# will be updated based on input matrices
		self.bPrettyFraction = formatFraction		# apply formatting to fractions - yes or no
		self.bShowBestResponse = showBestResponse		# show a box around the best response

		self.filename = fname
		self.f = open(self.filename)
		self.lines = self.f.readlines() 
		self.splitMatrices()
	
		# best response corresponding to a1
		self.br1 = [[None for i in range(self.numCols)] for j in range(self.numRows)]	
		# best response corresponding to a2		
		self.br2 = [[None for i in range(self.numCols)] for j in range(self.numRows)]	

		if self.bShowBestResponse and not self.bSinglePayoff:
			 self.calculateBestResponse()

	def splitMatrices(self):
		""" Parse input matrix, determine rows & cols, and place into two nested lists, a1, a2."""
		bSwitch = False
		for li in self.lines:
			if bSwitch: 
				self.a2.append(li.split())
			elif not li.strip():  #checking for blank lines as delimiter between 1st & 2nd matrix
				bSwitch = True
			else:
				self.a1.append(li.split())
		
		if not self.a2:
			self.a2 = self.a1
			self.bSinglePayoff = True
			self.bShowBestResponse = False

		self.numCols = len(self.a1[0]);
		self.numRows = len(self.a1)

	def getLaTeXString(self):
		""" Generate the contents of the LaTeX file in string form."""
		s = "\\documentclass{article}\n\\input{gamesty}%\n\\begin{document} \n"
		s = s + self.mapTokensToBMTemplate()
		return s + "\\end{document} \n"

	def mapTokensToBMTemplate(self):
		""" Create the LaTeX bimatrix macro inputs and wrap in appropriate macro format."""
		s ="\\bimatrixgame{4mm}"
		s = s + "{" + `self.numRows` + "}{" + `self.numCols` + "}{I}{II}%\n"
		
		#row labels & column labels
		s = s + self.generateRowLabels()
		s = s + self.generateColumnLabels()
		
		#payoff pairs
		s = s + self.generatePayoffPairs()

		return s

	def generateRowLabels(self): 
		""" Generate the labels for the row player strategy according to the defaults."""
		if self.numRows == 1:
			label = "{{M}}%\n"
		elif self.numRows == 2:		
			label = "{{T}{B}}%\n"
		elif self.numRows == 3:
			label = "{{T}{M}{B}}%\n"
		else:
			label = self.generateDefaultLabels(self.numRows)
	
		return label

	def generateColumnLabels(self): 
		""" Generate the labels for the column player strategy according to the defaults."""
		if self.numCols == 1:
			label = "{{c}}%\n"
		elif self.numCols == 2:		
			label = "{{l}{r}}%\n"
		elif self.numCols == 3:
			label = "{{l}{c}{r}}%\n"
		else:
			label = self.generateDefaultLabels(self.numCols)
	
		return label

	def generateDefaultLabels(self, numLabels):
		""" Generate the default row or column labels for matrices with > 3 rows/cols."""
		#97 = ascii 'a' - label cols a, b, c, d, e...etc.	
		label = "{"	
		for i in range(97, 97 + numLabels):
			label = label + "{" + chr(i) + "}"
		return label + "}%\n"
	
	def generatePayoffPairs(self): 	
		""" Generate the formatted payoff pairs values for all payoff pairs."""
		bBox = False
		s="{\n"
		r=""
		strPayoff = "\\payoffpairs{"
		
		if (self.bSinglePayoff):
			strPayoff = "\\singlepayoffs{"
		
		for i in range(0, self.numRows):
			r = strPayoff + `i+1` + "}{"
			
			for j in range(0, self.numCols):   	# player I payoffs
				if (self.bShowBestResponse):
					bBox = self.br1[i][j]
				r = r + self.wrapPayoff(self.a1[i][j], bBox)

			r = r + "}"
			
			if not self.bSinglePayoff:
				r = r + "{"
				
				for j in range(0, self.numCols): # player II payoffs
					if (self.bShowBestResponse):
						bBox = self.br2[i][j]
					r = r + self.wrapPayoff(self.a2[i][j], bBox)

				r = r + "}"

			s = s + r + "\n"
		
		return s + "}\n"

	def wrapPayoff(self, p, bBox):
		""" Generate the formatted payoff pair text for one set of payoff pairs."""
		# format at the individual payoff level
		if bBox: 
			lBox, rBox = "{\\fbox", "}"
		else:
			lBox, rBox = "", ""

		
		if not self.bPrettyFraction:		# pretty fractions OFF
			p = "{$" + p + "$}"
		elif p.count("/") > 0:		# reformat payoff for nicer fractions
			if p.count("-") == 1: 	# negative number
				negSign = "-"
			else:
				negSign = ""
			
			# split into numerator and denominator
			pFrac = p.split("/")
			pFrac[0] = pFrac[0].lstrip("-")
			pFrac[1] = pFrac[1].lstrip("-")
			
			p = "{$" + negSign + "\\frac{" + pFrac[0] + "}{" + pFrac[1] + "}$}"
	        else: 				# not a fraction
			p = "{$" + p + "$}"
		
		return lBox + p + rBox;


	def stringToDouble(self, s):
		""" Convert a string representing an integer or fraction to floating point representation."""
		#convert string to double
		#assumes either integer or fractional entries
		#does not handle entries with decimal point

		#int num;
		#int den;
		
		if s.count("/") == 0:
			n = float(s)
		else:
			sFrac = s.split("/")
	        	num = float(sFrac[0])
	        	den = float(sFrac[1])
	        	n = num/den

		return n	

	def calculateBestResponse(self):
		""" Calculate the payoffs that represent the best response for the row and column players."""
		n1 = [[0.0 for i in range(self.numCols)] for j in range(self.numRows)]
		n2 = [[0.0 for i in range(self.numCols)] for j in range(self.numRows)]

		
		rmax = [0.0 for i in range(self.numRows)]  		#row max values
		# double rtmp;
		
		cmax = [0.0 for i in range(self.numCols)] 		#col max values
		# double ctmp;
		
		#calculate best responses for first player
		for i in range(0, self.numRows):
			rmax[i] = self.stringToDouble(self.a2[i][0])

			for j in range(0, self.numCols):
				rtmp = self.stringToDouble(self.a2[i][j])
				n2[i][j] = rtmp

				if rmax[i] < rtmp:
					rmax[i] = rtmp

		# calculate best responses for second player
		for j in range(0, self.numCols):
			cmax[j] = self.stringToDouble(self.a1[0][j])

			for i in range(0, self.numRows):
				ctmp = self.stringToDouble(self.a1[i][j])
				n1[i][j] = ctmp
				
				if cmax[j] < ctmp:
					 cmax[j] = ctmp

		# store best response (T or F) in array for use in formatting table entry later
		for j in range(0, self.numCols):
			for i in range(0, self.numRows):
				if n2[i][j] < rmax[i]:
					self.br2[i][j] = False
				else:
					self.br2[i][j] = True
				
				if n1[i][j] < cmax[j]:
					self.br1[i][j] = False
				else:
					self.br1[i][j] = True
	
	def createLaTeXFile(self):
		f = open(self.filename + ".tex", 'r+')
		f.write(self.getLaTeXString())

def main(args): 

	options, rest = getopt.getopt(args, 'fbhn:', ['FractionFormatOff', 'BestResponseOff', 'Filename', 'help'])

	bFormatFraction = True
	bShowBestResponse = True

	helpString = "usage: \npython normalFormToLaTeX.py\n"
	helpString += " -n FILENAME [(req.) file contains 1 or 2 matrices, separated by a blank line]\n -f [(opt.) turn fraction format OFF]\n"
	helpString += " -b [(opt.) turn best response format OFF]"

	helpPrinted = False

	for opt, arg in options:
    		if opt in ('-f', '--FractionFormatOff'):
       			bFormatFraction = False
   		elif opt in ('-b', '--BestResponseOff'):
        		bShowBestResponse = False
		elif opt in ('-n', '--Filename'):
			print arg
			filename = arg
		elif opt in ('-h', '--help'):
        		print helpString
			helpPrinted = True
	
	try:
		a = normalFormToLaTeX(filename, bFormatFraction, bShowBestResponse)
		a.createLaTeXFile() 
		print "Created file " + filename + ".tex"
	except:
		if not helpPrinted:
			print helpString

if __name__ == '__main__': 
	main(sys.argv[1:]) 
